elastic/kibana
# [Uptime] implement encrypted saved objects for monitor management

## Summary

Resolves https://github.com/elastic/kibana/issues/127158

[Registers](https://github.com/elastic/kibana/pull/127348/files#diff-0e2159135205dee5b6cef3d83eebf4767f94b3a664d68887a2489c77d7ad6574R39-R51) the `syntheticsMonitor` saved object type as an encrypted saved object.

Defines a [set of sensitive keys](https://github.com/elastic/kibana/pull/127348/files#diff-6636c6ccff6a894ead9cb06f7526effc479e978ed070800f1333d209d0453f10R71-R90). 

Stores are sensitive fields on a single `secrets` key as a JSON payload.

[Formats](https://github.com/elastic/kibana/pull/127348/files#diff-a788779bd19101694ea64b27c6cbf2ce2225e3d98992dafa4f7fe737dd4f356cR41-R44) flattened `SyntheticsMonitor` type into a `SyntheticsMonitorsWithSecrets`, omitting all secrets from the resulting saved object and moving them onto the `secrets` field.

[Decrypts](https://github.com/elastic/kibana/pull/127348/files#diff-261a70a3f4310d15983c7e4ec151a88a39edc5611a6b78c870b1fcb6e0529d82R39) monitors from the `get_monitor` request, [flattening](https://github.com/elastic/kibana/pull/127348/files#diff-261a70a3f4310d15983c7e4ec151a88a39edc5611a6b78c870b1fcb6e0529d82R39) the `SyntheticsMonitorWithSecrets` type into a `SyntheticsMonitor`.

[Decrypts](https://github.com/elastic/kibana/pull/127348/files#diff-05f832378ee1977230af36dafea749010bcf43a4ca934543c335a43b207ecbe7R316-R326) the monitors from the sync task.

Ensures that monitors are able to be decrypted across multiple spaces in the [sync task](https://github.com/elastic/kibana/pull/127348/files#diff-05f832378ee1977230af36dafea749010bcf43a4ca934543c335a43b207ecbe7R322) and the [`get_monitor`](https://github.com/elastic/kibana/pull/127348/files#diff-fd99b8bcef8ddfd75f92d53fbbcdd3b574ecaadd4b5943d52e7c059d75105dc7R38) request.


### Testing

Testing should be conducted for monitors with sensitive fields, including HTTP, TCP, and Browser monitors.

Note: This branch is incompatible with previous saved objects. Testing should be done on a clean ES without monitor saved objects. You can delete your saved objects before testing this PR, or use a fresh ES instance via `yarn es snapshot` or `oblt-cli`.

Testing Decryption within the Client
---

1. Navigate to monitor management
2. Create a monitor. Ideally, create the monitor with all available sensitive fields for that monitor type. Review the list of [secret keys](https://github.com/elastic/kibana/pull/127348/files#diff-6636c6ccff6a894ead9cb06f7526effc479e978ed070800f1333d209d0453f10R71-R90) for reference. 
3. Ensure the monitor is loaded appropriately on the Monitor Management table.
4. Edit the monitor. Ensure that the sensitive fields are populated in the appropriate fields
5. Navigate to Stack Management -> Saved Objects. Filter saved object types by `synthetics-monitor`. Inspect your monitor saved object. Ensure that no sensitive fields from the list of [secret keys](https://github.com/elastic/kibana/pull/127348/files#diff-6636c6ccff6a894ead9cb06f7526effc479e978ed070800f1333d209d0453f10R71-R90) appears in the document.
6. Navigate to dev tools and run the following query. Ensure that the `secrets` key is encrypted.
```
GET .kibana*/_search
{
  "query": {
    "bool": {
      "filter": [
        {
          "term": {
            "type": "synthetics-monitor"
          }
        }
      ]
    }
  }
}
```

Testing Decryption within the push step
---
1. Create an HTTP monitor for your elasticsearch instance.
2. Ensure that the monitor appears down with status 401.
3. Edit the monitor, adding your elasticsearch the username and password.
4. Ensure that the monitor appears up with status 200 on the next check

Testing Editing from the Monitor Management Page
--
1. Disable your elasticsearch monitor. Ensure that the monitor is disabled.
2. Re-enable the monitor, ensure the monitor still appears as up with status 200 on the next check (this ensures that the secret fields are not lost when editing from the monitor management table)

Testing Spaces
---
Perform the same two tests within a different space by creating and accessing a new space via Spaces.
